File: haskell/src/Hangman.hs
module Hangman
(
    startGame
) where

import WordList
import Data.Char
import System.IO

startGame :: String -> IO ()
startGame wordListPath = do 
        randomWord <- getRandomWord wordListPath
        gameResult <- playGame randomWord (wordToUnderscore randomWord) "" 10
        if gameResult
            then putStrLn ("You gussed the word '" ++ randomWord ++ "' correctly")
            else putStrLn ("You did not guess '" ++ randomWord ++ "' correctly and are now hanged")

wordToUnderscore :: String -> String
wordToUnderscore randomWord = map (\char -> '_') randomWord 

playGame :: String -> String -> String -> Int -> IO Bool
playGame randomWord currentGuess usedLetters 0 = printStatus randomWord currentGuess usedLetters 0 >> return False
playGame randomWord currentGuess usedLetters guessesLeft = do
        printStatus randomWord currentGuess usedLetters guessesLeft
        guess <- getGuess
        -- Computes new lists for the letters used and the current guess
        let updatedCurrentGuess = updateGuess guess currentGuess randomWord
        let updatedUsedLetters = updateGuessed guess usedLetters 
        -- Checks if the player have guessed the word correctly
        if (updatedCurrentGuess == randomWord) then return True
            -- Checks if the player have used a guess or the random word contains the letter of if it is already tried
            else if ((elem guess randomWord) || (elem guess usedLetters))
                then playGame randomWord updatedCurrentGuess updatedUsedLetters guessesLeft
                else playGame randomWord updatedCurrentGuess updatedUsedLetters (guessesLeft-1)

-- We read an entire line and then extracts the last charecter, to prevent left over chars in the buffer
getGuess :: IO Char
getGuess = putStr "Please enter your next guess: " >> hFlush stdout >> getLine >>= 
    (\input -> if null input then getGuess else
        if isAlpha $ head input then return $ head input else getGuess)

updateGuess :: Char -> String -> String -> String
updateGuess _ _ [] = []
updateGuess guess (cg:cGuess) (rw:rWord)
    -- Both upper and lower case are checked as the word could be written with a combination
    | (rw == (toUpper guess)) = [rw] ++ (updateGuess guess cGuess rWord)
    | (rw == (toLower guess)) = [rw] ++ (updateGuess guess cGuess rWord)
    | otherwise = [cg] ++ (updateGuess guess cGuess rWord)

updateGuessed :: Char -> [Char] -> [Char]
updateGuessed guess usedLetters 
    | elem guess usedLetters = usedLetters 
    | otherwise = usedLetters ++ [' ', (toLower guess)]

printStatus :: String -> String -> String -> Int -> IO ()
printStatus randomWord currentGuess usedLetters guessesLeft = do
        putStrLn ("\nWord: " ++ currentGuess)
        putStrLn ("Tries Left: " ++ (show guessesLeft))
        putStrLn ("Guessed Letters: " ++ usedLetters)
        putStrLn "\nThe gallow\n\n"
        -- Draws a small ascii gallow based on the number of guesses the player has left
        case guessesLeft of 10 -> putStrLn "\n\n\n\n\n\n   \n"
                            9 -> putStrLn "\n\n\n\n\n   \n_______________\n\n"
                            8 -> putStrLn "\n\n\n  |\n  |\n  |\n________________\n\n"
                            7 -> putStrLn "  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n"
                            6 -> putStrLn "  |--         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n"
                            5 -> putStrLn "  |------         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n"
                            4 -> putStrLn "  |------         \n  |     |\n  |\n  |\n  |\n  |\n  |\n________________\n\n"
                            3 -> putStrLn "  |------         \n  |     |\n  |     0\n  |\n  |\n  |\n  |\n________________\n\n"
                            2 -> putStrLn "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |\n  |\n  |\n________________\n\n"
                            1 -> putStrLn "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |\n  |\n________________\n\n"
                            0 -> putStrLn "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |    / \\\n  |\n________________\n\n"
        -- The program has a tendency to write to output at interesting times if we do not force it
        hFlush stdout                                  

File: rust/src/hangman.rs
use std::io::{Write, BufRead};

/** Public Functions **/
pub fn start_game(random_word: String) {
    //Starts the game and prints a message based on the result
    if play_game(&random_word, 10) {
        println!("Congratulations, you guessed the word \"{}\" correctly.\n", random_word);
    } else {
        println!("You did not guess \"{}\" correctly, and have been hanged.\n", random_word);
    }
}

/** Private Functions **/
fn play_game(random_word: &str, tries_total: u8) -> bool {
    let mut tries_left = tries_total;
    let mut guessed_chars = String::new();
    let rwl = random_word.len();
    let mut user_guess = (0..rwl).map(|_| "_").collect::<String>();

    while tries_left > 0 {
        let guess = get_guess_from_user();
        let incorrect_guess = update_user_guess(random_word, &mut user_guess, guess);
        let new_char = update_guessed_chars(&mut guessed_chars, guess);

        if incorrect_guess && new_char {
            tries_left -= 1;
        }

        if user_guess == random_word {
            return true;
        }
        print_status(tries_left, &user_guess, &guessed_chars);
    }
    false
}

fn update_user_guess(random_word: &str, user_guess: &mut String, guess: char) -> bool {
    let chars: Vec<(char, char)> = random_word.chars().zip(user_guess.chars()).collect();
    user_guess.clear();
    let mut incorrect_guess = true;
    for (r, g) in chars {
        if r == guess {
            user_guess.push(r);
            incorrect_guess = false;
        } else {
            user_guess.push(g);
        }
    }
    incorrect_guess
}


fn update_guessed_chars(guessed_chars: &mut String, guess: char) -> bool {
    if ! guessed_chars.contains(guess) {
        guessed_chars.push(guess);
        guessed_chars.push(' ');
        true
    } else {
        false
    }
}


fn get_guess_from_user() -> char {
    let mut buffer = String::new();
    loop {
        print!("Please enter your next guess: ");
        ::std::io::stdout().flush().ok().expect("flush stdout error");
        let handle = ::std::io::stdin();
        let guess = handle.lock().read_line(&mut buffer).expect("read error");

        //Simple hack as the newline always will be included with a char
        if guess > 1 {
            return buffer.chars().next().unwrap().to_ascii_lowercase();
        }
        buffer.clear();
    }
}

fn print_status(guesses_left: u8, current_guess: &str, guessed_chars: &str) {
    println!("Word: {}", current_guess);
    println!("Tries Left: {}", guesses_left);
    println!("Guessed Letters: {}\n", guessed_chars);

    println!("The Gallow\n");

    match guesses_left {
        10 => println!("\n\n\n\n\n\n\n"),
         9 => println!("\n\n\n\n\n   \n_______________\n\n"),
         8 => println!("\n\n\n  |\n  |\n  |\n________________\n\n"),
         7 => println!("  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n"),
         6 => println!("  |--         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n"),
         5 => println!("  |------         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n"),
         4 => println!("  |------         \n  |     |\n  |\n  |\n  |\n  |\n  |\n________________\n\n"),
         3 => println!("  |------         \n  |     |\n  |     0\n  |\n  |\n  |\n  |\n________________\n\n"),
         2 => println!("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |\n  |\n  |\n________________\n\n"),
         1 => println!("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |\n  |\n________________\n\n"),
         _ => println!("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |    / \\\n  |\n________________\n\n")
    }
}